embassy-net/src/tcp.rs

```diff
@@ -359,6 +417,55 @@ impl<'d> TcpIo<'d> {
         .await
     }
 
+    async fn send<F, R>(&mut self, f: F) -> Result<R, Error>
+    where
+        F: FnOnce(&mut [u8]) -> (usize, R),
+    {
+        let mut f = Some(f);
+        poll_fn(move |cx| {
+            self.with_mut(|s, _| {
+                if s.send_queue() == s.send_capacity() {
```

Comment:
Same as recv, this needs more checks. Doc says "This function returns `Err(Error::Illegal)` if the transmit half of the connection is not open". This'll hang instead of erroring in the case where the transmit half is not open but the buffer is full.